### PR TITLE
feat: Heatmaps instrumentation

### PR DESCRIPTION
**Pull request by benjackwhite**
## Changes

Needs https://github.com/PostHog/posthog/pull/21598
Needs https://github.com/PostHog/posthog/pull/21487

Adds a new Heatmaps class that takes care of extra telemetry for improved heatmaps

## TODO:
- [ ] Don't track clicks from the toolbar

## Checklist
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers
- [ ] Accounted for backwards compatibility of any changes (no breaking changes in posthog-js!)

### CODE REVIEW

File: src/heatmaps.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    private _getProperties(e: MouseEvent, type: string): Properties {
+        // We need to know if the target element is fixed or not
+        // If fixed then we won't account for scrolling
+        // If not then we will account for scrolling
+
+        const scrollY = this.instance.scrollManager.scrollY()
+        const scrollX = this.instance.scrollManager.scrollX()
+        const scrollElement = this.instance.scrollManager.scrollElement()
+
+        const isFixedOrSticky = elementOrParentPositionMatches(e.target as Element, ['fixed', 'sticky'], scrollElement)
```
Comment (pauldambra) [inline src/heatmaps.ts:85]: I wonder if we could get computed styles to avoid walking the tree but 🤷

File: src/page-view.ts [hunk 2]
```diff
@@ -27,73 +14,49 @@ interface ScrollProperties {
...
-            prevPageViewData.maxScrollY == null ||
-            prevPageViewData.maxContentHeight == null ||
-            prevPageViewData.lastContentY == null ||
-            prevPageViewData.maxContentY == null
+            _isUndefined(maxScrollHeight) ||
+            _isUndefined(lastScrollY) ||
+            _isUndefined(maxScrollY) ||
+            _isUndefined(maxContentHeight) ||
+            _isUndefined(lastContentY) ||
+            _isUndefined(maxContentY)
```
Comment (pauldambra) [inline src/page-view.ts:55]: do we need to check `_isNullish` instead of `_isUndefined` or is null valid?

**Review by pauldambra — APPROVED** [review]
🚀

File: src/heatmaps.ts (hunk 1 above)
Comment (benjackwhite) [reply src/heatmaps.ts:85]: Sadly not. Computed styles do the whole inheritance tree but we actually don't want to know the current node's position, rather we want to know if it is in a fixed parent or not so fundamentally we have to traverse it.

**Pull request merged**